File: java/src/leetcode/editor/cn/P111_MinimumDepthOfBinaryTree.java
//给定一个二叉树，找出其最小深度。 
//
// 最小深度是从根节点到最近叶子节点的最短路径上的节点数量。 
//
// 说明：叶子节点是指没有子节点的节点。 
//
// 
//
// 示例 1： 
//
// 
//输入：root = [3,9,20,null,null,15,7]
//输出：2
// 
//
// 示例 2： 
//
// 
//输入：root = [2,null,3,null,4,null,5,null,6]
//输出：5
// 
//
// 
//
// 提示： 
//
// 
// 树中节点数的范围在 [0, 10⁵] 内 
// -1000 <= Node.val <= 1000 
// 
// Related Topics 树 深度优先搜索 广度优先搜索 二叉树 👍 784 👎 0

package leetcode.editor.cn.java;

import leetcode.editor.cn.java.DataStruct.TreeNode;

import java.util.ArrayDeque;

//Java：二叉树的最小深度
class P111_MinimumDepthOfBinaryTree {
    public static void main(String[] args) {
        Solution solution = new P111_MinimumDepthOfBinaryTree().new Solution();
    }
    //leetcode submit region begin(Prohibit modification and deletion)

    /**
     * Definition for a binary tree node.
     * public class TreeNode {
     * int val;
     * TreeNode left;
     * TreeNode right;
     * TreeNode() {}
     * TreeNode(int val) { this.val = val; }
     * TreeNode(int val, TreeNode left, TreeNode right) {
     * this.val = val;
     * this.left = left;
     * this.right = right;
     * }
     * }
     */
    class Solution {
        public int minDepth(TreeNode root) {
            if (root == null) {
                return 0;
            }
            ArrayDeque<TreeNode> arrayDeque = new ArrayDeque<>();
            arrayDeque.offer(root);
            int depth = 1;
            while (!arrayDeque.isEmpty()) {
                int len = arrayDeque.size();
                while (len > 0) {
                    TreeNode p = arrayDeque.poll();
                    if (p.left != null) {
                        arrayDeque.offer(p.left);
                    }
                    if (p.right != null) {
                        arrayDeque.offer(p.right);
                    }
                    if (p.left == null && p.right == null) {
                        return depth;
                    }
                    --len;
                }
                ++depth;
            }
            return depth;
        }
    }
//leetcode submit region end(Prohibit modification and deletion)

}

File: python/leetcode/editor/cn/P111_MinimumDepthOfBinaryTree.py
from collections import deque
from typing import Optional

from leetcode.editor.cn.dataStruct.TreeNode import TreeNode, createTree


# leetcode submit region begin(Prohibit modification and deletion)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        left = self.minDepth(root.left)
        right = self.minDepth(root.right)
        return min(left, right) + 1 if left and right else left + right + 1

    def minDepth2(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        res = 0
        cur = deque([root])
        while cur:
            for _ in range(len(cur)):
                node = cur.pop()
                res += 1
                if not node.left and not node.right:
                    return res
                if node.left:
                    cur.appendleft(node.left)
                if node.right:
                    cur.appendleft(node.right)


# leetcode submit region end(Prohibit modification and deletion)
print(Solution().minDepth2(createTree([3, 9, 20, None, None, 15, 7])))
print(Solution().minDepth2(createTree([2, None, 3, None, 4, None, 5, None, 6])))
